### ISSUE TITLE
Missuse of "authority" component

### ISSUE THREAD

**Issue by ndtodoroff** [timbray/fedi-uri#2]
I don't think the current scheme with special "authorities" to denote hashtags and search terms is a good idea. Relevant is [RFC 3986 section 3.2.2](https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.2), in particular the following two excerpts:

> The host subcomponent of authority is identified by an IP literal encapsulated within square brackets, an IPv4 address in dotted-decimal form, or a registered name. The host subcomponent is case-insensitive. The presence of a host subcomponent within a URI does not imply that the scheme requires access to the given host on the Internet. In many cases, the host syntax is used only for the sake of reusing the existing registration process created and deployed for DNS, thus obtaining a globally unique name without the cost of deploying another registry. However, such use comes with its own costs: domain name ownership may change over time for reasons not anticipated by the URI producer. In other cases, the data within the host component identifies a registered name that has nothing to do with an Internet host.  We use the name "host" for the ABNF rule because that is its most common purpose, not its only purpose.

>  A host identified by a registered name is a sequence of characters usually intended for lookup within a locally defined host or service name registry, though the URI's scheme-specific semantics may require that a specific registry (or fixed name table) be used instead.  The most common name registry mechanism is the Domain Name System (DNS).

"hashtag" and "search" could (as far as I can tell) reasonably be thought of as service names (though "service name" doesn't appear to be defined in this RFC), and the first excerpt explicity says that "authority" need not have anything to do with an internet host. However, I think mixing DNS lookup with special service names is unecessary and confusing. I am curious if there are any examples of existing URI schemes that do this.

My proposal would be to

  * Require `host` in `fedi://[user@]host` to go through DNS lookup when it is not an IP address, and
  * Require when an authority is not specified that the first segment of the path be a special identifier (in this case `hashtag` or `search`) and that the following segments are processed according to the special identifier. The examples `fedi://hashtag/fediverse` and `fedi://search/calico%20cat` would instead be `fedi:hashtag/fediverse` and `fedi:search/calico%20cat` under this proposal.
  * I also think it would be best to disallow absolute paths when no authority is present, so that `fedi:/hashtag/fediverse` would be invalid.

Another possibility would be to define a query syntax and use something like `fedi:?hashtag=fediverse`.

**Comment by timbray** [comment]
Thanks. Thought-provoking. I'll have to go look at 3986 some more . In your proposal, the "//" after the scheme becomes the signifier that what follows contains an Authority; is there backing in some RFC for that being a principled position?

I guess another option would be fedi://?hashtag=fediverse, then the spec would say "the Authority and Path MUST be omitted if the Query is provided". I confess that my motivation is just that some fedi URIs having // and others not looks kind of jarring.

**Comment by ndtodoroff** [comment]
@timbray Yes, that is how the syntax is defined. From the RFC, section 3:
```
      URI         = scheme ":" hier-part [ "?" query ] [ "#" fragment ]

      hier-part   = "//" authority path-abempty
                  / path-absolute
                  / path-rootless
                  / path-empty
```

The unquoted slash means "or" and the unquoted brackets mean "optional". Quoted strings mean the contents of the quotes must be present literally. They give the following two examples:

```
         foo://example.com:8042/over/there?name=ferret#nose
         \_/   \______________/\_________/ \_________/ \__/
          |           |            |            |        |
       scheme     authority       path        query   fragment
          |   _____________________|__
         / \ /                        \
         urn:example:animal:ferret:nose
```
A URI like `fedi://?query` specifies a default host insofar as

> If the URI scheme defines a default for host, then that default applies when the host subcomponent is undefined or when the registered name is empty (zero length).

The RFC doesn't seem to specify any meaning when the scheme does not define a default host, but in my opinion the intent seems to be for an empty host to indicated a default host, so I would push back against a syntax like `fedi://?query`.

**Comment by timbray** [comment]
Yeah… I've been around the track at the IETF a few times, you're right about what the ABNF says.

But… going back to your very first point, I read that language from 3986 3.2.2, and I think that "hashtag" and "search" and so on fall very nicely into 3986's notion of a "registered name" - it points out that the DNS is conventionally used, but doesn't need to be - at the cost of maintaining another registry.  Well, if we decided to follow the RFC route, it's easy enough for an RFC to create an IANA registry, we just did that for JSONPath functions - I co-chaired that WG and the IESG approved our output, it'll have an RFC number in January I think. There's a little bit of bureaucracy but surprisingly little. So IF there's User Information after the //, which is the case if a "@" is present, then it's a DNS name, otherwise it's the IANA Fedi Authority Registry.

**Comment by ndtodoroff** [comment]
I don't think you're wrong. But it feels overcomplicated to me. I like the clear syntactic distinction that `fedi://user@host` and `fedi:query/term` gives, even more so if we allow `fedi://host` to represent any Fediverse-compatible host without reference to any user.

Another idea I had seems incompatible with the `fedi://query`
syntax. We could have `fedi:search/term` mean "search the whole Fediverse" (as best the application consuming the URI can do, or perhaps restricted by some user-specified list of hosts) but have `fedi://host/search/term` mean "search this host specifically". This extends nicely to posts/notes/toots (or anything really) where if there were some global way of identifying a note then we could have the following:

  * `fedi:note/id` specifies the note with that ID, but not where to find it. An application that consumes this URI might have some list of known hosts it can query to try to find the note.
  * `fedi://host/note/id` specifies the note with that ID, and that note can be found at `host`.

**Comment by hardie** [comment]
I think one question here is how many actions you want the URI scheme to support.  I can see three fairly obvious ones:

follow (a user or thread)
block (a user or server)
search (for a hashtag or other data)

If you want to support all of them, then a syntax which distinguishes among the different methods would be useful, and it would be useful for it to be common across the methods.
